tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py

```diff
@@ -331,3 +332,147 @@ def test_find_unification_and_try_specializing_pnad():
 
     learner.recompute_datastores_from_segments([new_pnad])
     assert len(new_pnad.datastore) == 1
+
+
+def test_keep_effect_data_partitioning():
+    """Test that the BackchainingSTRIPSLearner is able to correctly induce
+    operators with keep effects in a case where a naive procedure that always
+    induces potential keep effects would fail."""
+
+    utils.reset_config({"segmenter": "atom_changes"})
```

Comment:
high-level idea:
two demos: [TurnOn, Configure, Run] and [Configure, TurnOn, Run]
TurnOn always just turns on your machine, Configure makes it configured but may turn on/off arbitrary other machines. you are allowed to Configure when the machine is either off or on.
our algorithm will say that MachineOn is a keep effect of Configure, but then this will be harmful to the second demo, where the machine was off when you configured it. so you need to keep around the original (non-KEEP) operator for Configure. see ground truth operators at end of function

Reply:
add this as a comment?

Reply:
to clarify,
>you are allowed to Configure when the machine is either off or on.

but only if `MachineConfigurableWhileOff` holds?

Reply:
yup that's right, that precond we just added to test that the original operator (the Configure without keep effects) needs its preconditions re-induced, since they can get more specific now that you've thrown the keep effect data into the keep effect operator.